### PR TITLE
CNTRLPLANE-3565: test and fix destroy grace period lifecycle

### PR DESCRIPTION
**Pull request by mgencur**
## What this PR does / why we need it:

Add TestDestroyGracePeriod with 8 test cases covering annotation parsing, condition creation, requeue timing, and finalizer removal.

Fix two bugs found during testing:
- Grace period condition check now handles ConditionFalse status
- Use injected r.Clock instead of bare time.Now() for testability

## Which issue(s) this PR fixes:

Fixes https://redhat.atlassian.net/browse/CNTRLPLANE-3565

## Checklist:
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [x] This change includes unit tests.

## Release Notes

* **Bug Fixes**
  * Improved HostedCluster deletion grace-period handling by switching to the reconciler-injected deterministic clock for timing and requeue calculations.
  * When a grace period is set, the `HostedClusterDestroyed` condition is initialized or corrected (when missing or not `True`) with an updated `LastTransitionTime`, and the controller requeues only for the remaining grace duration.

* **Tests**
  * Added table-driven unit test coverage for destroy grace-period behavior, including validation errors, condition state transitions, requeue timing, and finalizer removal outcomes.

### CODE REVIEW

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go [hunk 1]
```diff
@@ -7761,3 +7761,245 @@ func TestPropagateAzureResourceIDAnnotation(t *testing.T) {
...
+	}{
+		{
+			name:            "When no grace period annotation is set, it should remove the finalizer immediately",
+			expectFinalizer: false,
+		},
+		{
+			name:                  "When grace period annotation is invalid, it should return an error",
+			gracePeriodAnnotation: "not-a-duration",
+			expectError:           true,
+			expectFinalizer:       true,
```
Comment (mgencur) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go:8897]: This test currently passes with `expectFinalizer: true` but this behaviour is strange:

With an invalid annotation, time.ParseDuration fails at line 484 of hostedcluster_controller.go and the reconciler returns an error before it ever reaches the deletion logic:
```
  hcDestroyGracePeriod, err = time.ParseDuration(gracePeriodString)
  if err != nil {
      return ctrl.Result{}, fmt.Errorf("failed to parse %s annotation: %w", hyperv1.HCDestroyGracePeriodAnnotation,
  err)
  }
```
This early return means delete() never runs, the finalizer is never removed, and the HC stays in a stuck-deleting state. The reconciler keeps retrying via error-based requeue, waiting for someone to fix the annotation. 

I would expect, with invalid annotation the finalizer would be removed because it was a mistake and we don't want to hang forever in that case.

@jparrill  WDYT ?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 2]
```diff
@@ -588,9 +588,9 @@ func (r *HostedClusterReconciler) reconcile(ctx context.Context, req ctrl.Reques
 				return ctrl.Result{RequeueAfter: hcDestroyGracePeriod}, nil
 			}
 
-			if time.Since(hostedClusterDestroyedCondition.LastTransitionTime.Time) < hcDestroyGracePeriod {
+			if r.Clock.Since(hostedClusterDestroyedCondition.LastTransitionTime.Time) < hcDestroyGracePeriod {
 				log.Info("Waiting for grace period", "gracePeriod", hcDestroyGracePeriod)
-				return ctrl.Result{RequeueAfter: hcDestroyGracePeriod - time.Since(hostedClusterDestroyedCondition.LastTransitionTime.Time)}, nil
+				return ctrl.Result{RequeueAfter: hcDestroyGracePeriod - r.Clock.Since(hostedClusterDestroyedCondition.LastTransitionTime.Time)}, nil
```
[the remark this answers is not in the document]
Comment (mgencur) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go]: Fixed.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 3]
```diff
@@ -579,13 +579,13 @@ func (r *HostedClusterReconciler) reconcile(ctx context.Context, req ctrl.Reques
 		}
 
 		if hcDestroyGracePeriod > 0 {
-			if hostedClusterDestroyedCondition == nil {
+			if hostedClusterDestroyedCondition == nil || hostedClusterDestroyedCondition.Status != metav1.ConditionTrue {
```
Comment (jparrill) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:582]: Without this, if the condition existed with `Status=False`, the code skipped the re-creation block and fell straight into the elapsed-time check with a stale `LastTransitionTime` — grace period could expire instantly.

`reconcile_legacy.go` has the same unfixed pattern (~lines 223-243). The deletion guard at line 148 already has `|| Status != ConditionTrue`, but the grace period block doesn't. Intentional or did it slip through?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 4]
```diff
@@ -597,9 +597,10 @@ func (r *HostedClusterReconciler) reconcile(ctx context.Context, req ctrl.Reques
 				return ctrl.Result{RequeueAfter: hcDestroyGracePeriod}, nil
 			}
 
-			if time.Since(hostedClusterDestroyedCondition.LastTransitionTime.Time) < hcDestroyGracePeriod {
+			elapsed := r.Clock.Since(hostedClusterDestroyedCondition.LastTransitionTime.Time)
```
Comment (jparrill) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:600]: The old code called `time.Since()` twice — untestable and subtly racy (two calls returning slightly different values). Extracting `elapsed` fixes both.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go [hunk 5]
```diff
@@ -8832,3 +8832,245 @@ func TestReconcileSREMetricsConfig_EffectiveMetricsSet(t *testing.T) {
...
+
+	tests := []struct {
+		name                  string
+		gracePeriodAnnotation string
+		existingCondition     *metav1.Condition
+		expectError           bool
+		expectFinalizer       bool
+		expectCondition       bool
+		expectConditionReason string
+		expectConditionStatus metav1.ConditionStatus
```
Comment (jparrill) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go:8885]: `expectConditionStatus` is declared in the struct but never assigned in any of the 8 cases — always falls back to `ConditionTrue` via zero value. Either remove it or add a case that actually uses it (e.g. `ConditionUnknown` to confirm the `!= ConditionTrue` guard handles all three valid K8s states).

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go [hunk 6]
```diff
@@ -8832,3 +8832,245 @@ func TestReconcileSREMetricsConfig_EffectiveMetricsSet(t *testing.T) {
 		})
 	}
 }
+
+func TestDestroyGracePeriod(t *testing.T) {
```
Comment (jparrill) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go:8836]: The guard checks `!= ConditionTrue`, which handles `ConditionUnknown` too (valid K8s status), but there's no test for it. Is that intentional or just untested? If it's by design, a quick variant like test case #7 but with `ConditionUnknown` would document the intent.

**Review by jparrill — COMMENTED** [review]
Dropped some comments. Thanks!

Main ask: confirm whether `reconcile_legacy.go` needs the same fixes (lines ~223-243). It has the same `time.Now()` / `time.Since()` and the same incomplete guard on the grace period block.

Test nits (unused `expectConditionStatus` field, missing boundary case) are non-blocking.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 3 above)
Comment (mgencur) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:582]: It slipped. I will fix it too.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 4 above)
Comment (mgencur) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:600]: I think this is already fixed in this PR by extracting "elapsed". If not, please let me know.

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go (hunk 5 above)
Comment (mgencur) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go:8885]: Added a testcase

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go (hunk 6 above)
Comment (mgencur) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go:8836]: Added the test case

**Review by jparrill — APPROVED** [review]
All 4 comments addressed — legacy reconciler fixed, elapsed extraction confirmed, ConditionUnknown test added, expectConditionStatus field now exercised. LGTM.

/approve

**Review by mehabhalodiya — APPROVED** [review]
/lgtm
